woocommerce/woocommerce-ios
# Dashboard stats: replace XLPagerTabStrip with WPiOS implementation

Part of #1911 


## Why

In the dashboard, we currently use a third-party library `XLPagerTabStrip` for the tabbed stats where there is a tab bar for selecting the time range with scrollable stats below it. There have been discussions of removing `XLPagerTabStrip` from our app as linked in #1911, and we reached a conclusion to replace it with the WPiOS `FilterTabBar` implementation in p91TBi-47N-p2 after design confirmation. This used to be a tech debt item left for the future. Now that we are moving to one single vertical `UIScrollView` to account for the dynamic onboarding card height, the use of `XLPagerTabStrip` causes a bunch of issues beyond our control from my testing. Replacing `XLPagerTabStrip` enables us to make changes that suit our design. [The library](https://github.com/xmartlabs/XLPagerTabStrip) has also accumulated 250+ open issues and could go out of maintenance anytime. Since WPiOS has been using the `FilterTabBar` + `TabbedViewController` implementation for a few years, we have a pretty stable alternative to replace with.

## Description


Apology for the big diffs, the new diffs are mostly from the WPiOS [`FilterTabBar`](https://github.com/wordpress-mobile/WordPress-iOS/blob/c67f3b6205c1bc40085b390ec5e46faf5e281df9/WordPress/Classes/ViewRelated/System/FilterTabBar.swift) and  [`TabbedViewController`](https://github.com/wordpress-mobile/WordPress-iOS/blob/c67f3b6205c1bc40085b390ec5e46faf5e281df9/WordPress/Classes/ViewRelated/Reader/Manage/TabbedViewController.swift) with some updates to fit our use case like [updates to make the build pass](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/d01d6b23bb2c677a7ec790d591ce3ae2f0b84589) and [programmatically selecting a time range tab](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/c68b19c4b1c8a60c48dcd5fd3a8c1f5728f2b65e).

Other than the new classes adapted from WPiOS, the majority of the diffs are on `StoreStatsAndTopPerformersViewController` where it's now a subclass of `TabbedViewController` instead of `XLPagerTabStrip.ButtonBarPagerTabStripViewController`. `selection: Int` from `TabbedViewController` is the source of truth of the selected tab index as shown in the UI, and I kept the observable version of it `selectedTimeRangeIndex` to minimize the changes. I spent quite some time [updating the tab bar button styles](https://github.com/woocommerce/woocommerce-ios/pull/8942/commits/ed794334f81c9fd9849aaf7adb8d3f2aadfb7824) in order to remove the iOS 15+ build warnings on `contentEdgeInsets` because the new button configuration breaks certain button style customizations with the old APIs.

⚠️ One style that I couldn't get exactly right is the spacing of the tab bar buttons. When setting the `UIStackView`'s `spacing` to `0` and `distribution` to `euqalSpacing`, a ~16px spacing is still shown. From the [Apple documentation](https://developer.apple.com/documentation/uikit/uistackview/1616225-spacing) on `UIStackView.spacing`, it's the minimum spacing but I ran out of ideas on how to set the actual spacing and I set a negative value as a workaround. Please feel free to share any ideas on this!

## Testing instructions


- Log in if needed --> the stats should look like before, with the same functionalities except for horizontal scrolling between time range tabs
- Rotate the device --> the dashboard should work as before in the landscape mode
- Tap another time range tab --> the corresponding time range stats should be shown
- Relaunch the app --> the last selected tab and stats should be shown
- Switch to another store --> the stats for the new store should be shown

## Screenshots


\ | Before | After 
-- | -- | --
light | ![Simulator Screen Shot - iPhone 14 Pro - 2023-02-23 at 16 09 51](https://user-images.githubusercontent.com/1945542/220853240-ccd42a23-de5d-4d51-ae03-7e7d9ba01e58.png) | ![Simulator Screen Shot - iPhone 14 - 2023-02-23 at 16 09 53](https://user-images.githubusercontent.com/1945542/220853231-00e19ddc-aaeb-4f54-9b2a-51cdbd3a39b6.png)
dark | ![Simulator Screen Shot - iPhone 14 Pro - 2023-02-23 at 16 11 31](https://user-images.githubusercontent.com/1945542/220853299-466dd393-be5c-4304-80a5-7c97260dbc67.png) | ![Simulator Screen Shot - iPhone 14 - 2023-02-23 at 16 11 33](https://user-images.githubusercontent.com/1945542/220853290-e3d63179-5ec5-496a-ab56-1f81ef41fd2b.png)


---
